### PR TITLE
Generate client API for trace CRD

### PR DESCRIPTION
**Pull request by mauriciovasquezbernal**
Use client-gen to generate a client library for handling the trace resources. 

TODO:
- [x] Understand why the generated code requires a change to compile
- [x] Update controller-gen script? 
- [x] Update CI to detect outdated code 
- [x] How to handle `go mod vendor` that is needed to get the contrib-update script?

### CODE REVIEW

**Comment by mauriciovasquezbernal** [comment]
It's ready for review! I decided not to use the watch interface (as done in https://github.com/kinvolk/inspektor-gadget/pull/398). I got it to work but then I realized there are so many considerations we need to discuss, so I prefer to to this change in https://github.com/kinvolk/inspektor-gadget/pull/398.

File: pkg/client/clientset/versioned/typed/gadget/v1alpha1/trace.go [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+	DeleteCollection(ctx context.Context, opts v1.DeleteOptions, listOpts v1.ListOptions) error
+	Get(ctx context.Context, name string, opts v1.GetOptions) (*v1alpha1.Trace, error)
+	List(ctx context.Context, opts v1.ListOptions) (*v1alpha1.TraceList, error)
+	Watch(ctx context.Context, opts v1.ListOptions) (watch.Interface, error)
+	Patch(ctx context.Context, name string, pt types.PatchType, data []byte, opts v1.PatchOptions, subresources ...string) (result *v1alpha1.Trace, err error)
+	TraceExpansion
+}
+
+// traces implements TraceInterface
+type traces struct {
```
Comment (eiffel-fl) [inline pkg/client/clientset/versioned/typed/gadget/v1alpha1/trace.go:52]: I am not sure if this code is generated, but why `traces` instead of `trace`?

**Comment by eiffel-fl** [comment]
Hi.


Thank you for it, it makes the code easier to read and I like the fact the last commit has more removed lines than added ones!
I tested it running `biolatency` on `minikube` because this gadget triggers the most of your modifications and everything was OK.
So, LGTM and: 
Tested-by: Francis Laniel <flaniel at linux dot microsoft dot com>.


Best regards.

**Review by eiffel-fl — APPROVED** [review]

File: pkg/client/clientset/versioned/typed/gadget/v1alpha1/trace.go (hunk 1 above)
Comment (mauriciovasquezbernal) [reply pkg/client/clientset/versioned/typed/gadget/v1alpha1/trace.go:52]: I don't know the specific answer but those APIs are named with the plural.

```
traceClient.GadgetV1alpha1().Traces("gadget")
clientset.CoreV1().Pods("default")
```

etc.

Comment (eiffel-fl) [reply pkg/client/clientset/versioned/typed/gadget/v1alpha1/trace.go:52]: If this code is generated, I am OK with this, as you do not have a lot of control on generated code.
More generally, can you please indicate (in a comment here) which parts of the code you write and which were generated?

Comment (mauriciovasquezbernal) [reply pkg/client/clientset/versioned/typed/gadget/v1alpha1/trace.go:52]: Sure, everything under pkg/client was generated. Btw, all generated files have a `Code generated by client-gen` comment after the license header.

Comment (eiffel-fl) [reply pkg/client/clientset/versioned/typed/gadget/v1alpha1/trace.go:52]: Oh! I missed this comments :sweat_smile:.
Thank you for the precision!

File: cmd/kubectl-gadget/utils/trace.go [hunk 2]
```diff
@@ -132,42 +131,31 @@ func printTraceFeedback(m map[string]string, totalNodes int) {
...
-	traceConfig.APIPath = "/apis"
-	traceConfig.NegotiatedSerializer = serializer.NewCodecFactory(scheme.Scheme)
-	traceConfig.UserAgent = restclient.DefaultKubernetesUserAgent()
-
-	traceRestClient, err := restclient.UnversionedRESTClientFor(&traceConfig)
+	traceClient, err := clientset.NewForConfig(config)
 	if err != nil {
-		return nil, fmt.Errorf("Error setting up trace REST client: %w", err)
+		return nil, err
 	}
```
Comment (blanquicet) [inline cmd/kubectl-gadget/utils/trace.go:156]: Why did you remove the error message this function was appending to the returned errors? Are they good enough to understand what happened? I am saying that because all the `getTraceClient` callers are forwarding the error as it is. Therefore, the error returned by this function will be directly shown to the user.

File: cmd/kubectl-gadget/utils/trace.go [hunk 3]
```diff
@@ -200,18 +188,14 @@ func createTraces(trace *gadgetv1alpha1.Trace) error {
...
-			Resource("traces").
-			Body(trace).
-			Do(context.TODO()).
-			Error()
+		newtrace, err := traceClient.GadgetV1alpha1().Traces("gadget").Create(
+			context.TODO(), trace, metav1.CreateOptions{},
+		)
 		if err != nil {
-			traceID, present := trace.ObjectMeta.Labels[GLOBAL_TRACE_ID]
+			traceID, present := newtrace.ObjectMeta.Labels[GLOBAL_TRACE_ID]
```
Comment (blanquicet) [inline cmd/kubectl-gadget/utils/trace.go:195]: If `err != nil`, I don't think we can rely on `newtrace` being a valid pointer. I think we should still use `trace` here.

File: cmd/kubectl-gadget/utils/trace.go [hunk 4]
```diff
@@ -406,7 +376,7 @@ func SetTraceOperation(traceID string, operation string) error {
 
 // waitForTraceState loops over all trace whom ID is given as parameter
 // waiting until they are in the expected state.
-func waitForTraceState(traceID string, expectedState string) (gadgetv1alpha1.TraceList, error) {
+func waitForTraceState(traceID string, expectedState string) (*gadgetv1alpha1.TraceList, error) {
```
Comment (blanquicet) [inline cmd/kubectl-gadget/utils/trace.go:379]: I like that you took advantage of this PR to start using the pointer. Great move!

**Review by blanquicet — COMMENTED** [review]
I only reviewed the last commit where you used the generated client API. I think it is a great improvement!

I didn't find regressions during tests with biolatency and seccomp-advisor so, apart from the couple of comments I added, I think the PR is ready.

File: cmd/kubectl-gadget/utils/trace.go (hunk 2 above)
Comment (mauriciovasquezbernal) [reply cmd/kubectl-gadget/utils/trace.go:156]: Right. I added them back.

File: cmd/kubectl-gadget/utils/trace.go (hunk 3 above)
Comment (mauriciovasquezbernal) [reply cmd/kubectl-gadget/utils/trace.go:195]: Good catch!
